File: Data Structures & Algorithms/valid-tree/submission-4.js
class Solution {
    /**
     * @param {number} n
     * @param {number[][]} edges
     * @returns {boolean}
     */
    validTree(n, edges) {
        if(edges.length != n - 1) return false;

        const adj = Array.from({ length: n }, () => []);
        for(const [u, v] of edges) {
            adj[u].push(v);
            adj[v].push(u);
        }

        const vis = new Array(n).fill(false);
        vis[0] = true;

        const dfs = (u) => {
            for(const v of adj[u]) {
                if(!vis[v]) {
                    vis[v] = true;
                    dfs(v);
                }
            }
        }

        dfs(0);

        // for(const b of vis) if(!b) return b;
        // return true;
        // return vis.every(Boolean);
        return !vis.includes(false);
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-4.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = [[] for _ in range(n)]
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        seen = set()
        def dfs(i):
            if i in seen: 
                return 0
            seen.add(i)
            for j in adj[i]: 
                dfs(j)
            return 1

        return sum(dfs(i) for i in range(n))
        
        
        # adj = [[] for _ in range(n)]
        # vis = [False] * n

        # for u, v in edges:
        #     adj[u].append(v)
        #     adj[v].append(u)
        
        # def dfs(u: int):
        #     vis[u] = True
        #     for v in adj[u]:
        #         if not vis[v]:
        #             dfs(v)
        
        # ans = 0
        # for i in range(n):
        #     if not vis[i]:
        #         ans += 1
        #         dfs(i)
        
        # return ans